### PR TITLE
Event Streaming

### PR DESCRIPTION
**Pull request by quartzjer**
Streaming events is more better, go go application/jsonstream!

### CODE REVIEW

File: Collections/Contacts/package.json [hunk 1]
```diff
@@ -20,39 +20,48 @@
     "events": [
       [
         "contact/foursquare",
-        "/events"
+        "/events",
+        true
```
Comment (kristjan) [inline Collections/Contacts/package.json:24]: These are getting long and `true` doesn't mean anything unless you know the format. Maybe we want to switch to an object?

```
"events": [
  {
    "name" : "contact/foursquare",
    "path" : "/events",
    "stream" : true
  },
  // ...
]
```

File: Collections/Links/dataIn.js [hunk 2]
```diff
@@ -88,9 +88,9 @@ function getLinks(getter, lurl, callback) {
 function processEncounter(e, cb)
 {
     dataStore.enqueue(e, function() {
+        cb(); // return after we know it's queued
         encounterQueue.push(e, function(arg){
-            logger.verbose("QUEUE SIZE: "+encounterQueue.length());
-            cb();
+            logger.verbose("QUEUE SIZe: "+encounterQueue.length());
```
Comment (kristjan) [inline Collections/Links/dataIn.js:93]: "SIZe"

File: Collections/Links/dataIn.js [hunk 3]
```diff
@@ -88,9 +88,9 @@ function getLinks(getter, lurl, callback) {
 function processEncounter(e, cb)
 {
     dataStore.enqueue(e, function() {
+        cb(); // return after we know it's queued
```
Comment (kristjan) [inline Collections/Links/dataIn.js:91]: Why don't we want to wait anymore until the encounters are also queued?

File: Collections/Links/links.js [hunk 4]
```diff
@@ -47,17 +50,11 @@ app.get('/embed', function(req, res) {
...
-    }
-
-    // handle asyncadilly
-    dataIn.processEvent(req.body);
-    res.writeHead(200);
-    res.end('ok');
+  var q = async.queue(dataIn.processEvent, 1);
+  req.jsonStream(q.push, function(error){
+      if(error) logger.error(error);
+      res.send(200);
```
Comment (kristjan) [inline Collections/Links/links.js:56]: `200` even if there's an error?

File: Collections/Photos/photos.js [hunk 5]
```diff
@@ -26,18 +29,10 @@ app.get('/update', function(req, res) {
...
-        if (err) {
-            logger.error("Error processing: " + err);
-            return res.send(err, 500);
-        }
-
-        res.send("Event Handled");
+    var q = async.queue(dataIn.addEvent, 1);
+    req.jsonStream(q.push, function(error){
+        if(error) console.error(error);
+        res.send(200);
```
Comment (kristjan) [inline Collections/Photos/photos.js:35]: A lot of this stuff looks the same; can we put it somewhere common?

File: Collections/Contacts/package.json (hunk 1 above)
Comment (kristjan) [inline Collections/Contacts/package.json:24]: Or it looks from `levents` like the third parameter is `batching`. Definitely could use clarification :)

File: Common/node/levents.js [hunk 6]
```diff
@@ -124,3 +128,44 @@ function lqueue(lurl)
...
+    sendCb();
+  }, function() {
+    req.end();
+  });
+};
+function batchqueue(lurl) {
+  if (lqueues[lurl]) return lqueues[lurl];
+  lqueues[lurl] = new BatchSendQueue(lurl);
+  return lqueues[lurl];
+}
```
Comment (kristjan) [inline Common/node/levents.js:171]: `BatchSendQueue` looks worth
- pulling into another module
- testing

File: Collections/Contacts/package.json (hunk 1 above)
Comment (quartzjer) [inline Collections/Contacts/package.json:24]: Yeah, this all will go away once I get a chance to convert the collection data processing to something more like how synclets work, they're registered as exported callbacks, for now it's just a transient old format

File: Collections/Links/dataIn.js (hunk 3 above)
Comment (quartzjer) [inline Collections/Links/dataIn.js:91]: the expectation of the event processing from core is just that the collection has "got it", and that happens once it's in the mongo queue, the actual link processing is decoupled

File: Common/node/lservicemanager.js [hunk 7]
```diff
@@ -216,7 +216,8 @@ exports.mapReload = function(id)
     if(js.events) {
         for (var i = 0; i < js.events.length; i++) {
             var ev = js.events[i];
-            levents.addListener(ev[0], js.id, ev[1]);
+            var batching = (ev.length > 2 && ev[2] === true) ? true : false;
```
Comment (kristjan) [inline Common/node/lservicemanager.js:219]: Can't this be

```
var batching = (ev.length > 2 && ev[2] === true)
```

Or if we dont' want the short-circuit return value

```
var batching = !!(...)
```

File: Ops/webservice.js [hunk 8]
```diff
@@ -402,7 +402,9 @@ locker.get('/core/:svcId/listen', function(req, res) {
         return;
     }
     if(cb.substr(0,1) != "/") cb = '/'+cb; // ensure it's a root path
-    levents.addListener(type, svcId, cb);
+    var batching = false;
+    if (req.param("batch") === "true" || req.param === true) batching = true;
```
Comment (kristjan) [inline Ops/webservice.js:406]: Perhaps

```
var batching = (req.param("batch") === "true" || req.param === true)
```

File: package.json [hunk 9]
```diff
@@ -17,6 +17,7 @@
         "npm":"=1.1.1",
         "connect": "=1.8.5",
         "express": "=2.5.4",
+        "express-jsonstream": "=0.0.2",
```
Comment (kristjan) [inline package.json:20]: Some of these version numbers are unsettlingly low :-/

File: Collections/Links/dataIn.js (hunk 2 above)
Comment (quartzjer) [inline Collections/Links/dataIn.js:93]: SIZe

File: Common/node/levents.js [hunk 10]
```diff
@@ -124,3 +128,44 @@ function lqueue(lurl)
...
+BatchSendQueue.prototype.run = function() {
+  if (this.running) return;
+  if (this.items.length == 0) return;
+
+  this.running = true;
+  var sendingItems = this.items;
+  this.items = [];
+  logger.verbose("Sending " + sendingItems.length + " batched events to " + this.url);
+  var self = this;
+  var req = request({url:this.url, method:"POST", headers:{"content-type":"application/jsonstream"}}, function(err, res, body) {
```
Comment (smurthas) [inline Common/node/levents.js:151]: FYI, I wrote a module to wrap this up a bit more: https://github.com/smurthas/Request-JSONStream/blob/master/test.js#L30

No need to change now, but in the future.

File: Collections/Links/links.js (hunk 4 above)
Comment (quartzjer) [inline Collections/Links/links.js:56]: yes the event sender in core cannot do anything with errors, all we need to do here is close the socket

File: Collections/Photos/photos.js (hunk 5 above)
Comment (quartzjer) [inline Collections/Photos/photos.js:35]: @smurthas is very eager to yes, it's just lower priority as it's more of a cleanup than any functional improvement

**Pull request merged**